File: py/minsk/analysis/text/source.py
from dataclasses import dataclass

from minsk.analysis.text.span import TextSpan


@dataclass(frozen=True)
class TextLine:
    _text: "SourceText"
    start: int
    length: int
    length_including_line_break: int

    @property
    def end(self) -> int:
        return self.start + self.length

    @property
    def end_including_line_break(self) -> int:
        return self.start + self.length_including_line_break

    @property
    def span(self) -> TextSpan:
        return TextSpan(self.start, self.length)

    @property
    def span_including_line_break(self) -> TextSpan:
        return TextSpan(self.start, self.length_including_line_break)


class SourceText:
    _text: str
    lines: tuple[TextLine, ...]

    def __init__(self, text: str):
        self._text = text
        self.lines = self._parse_lines(text)

    def _parse_lines(self, text: str) -> tuple[TextLine, ...]:
        result = []

        position = 0
        line_start = 0

        while position < len(text):
            line_break_width = self._get_line_break_width(text, position)

            if line_break_width > 0:
                self._add_line(result, line_start, position, line_break_width)
                position += line_break_width
                line_start = position
            else:
                position += 1

        self._add_line(result, line_start, position, 0)

        return tuple(result)

    @staticmethod
    def _get_line_break_width(text: str, position: int) -> int:
        look = text[position + 1] if position + 1 < len(text) else "\0"
        match text[position]:
            case "\r" if look == "\n":
                return 2
            case "\r" | "\n":
                return 1
            case _:
                return 0

    def _add_line(
        self,
        result: list[TextLine],
        line_start: int,
        position: int,
        line_break_width: int,
    ):
        length = position - line_start
        length_including_line_break = length + line_break_width
        line = TextLine(self, line_start, length, length_including_line_break)
        result.append(line)

    def get_line_index(self, position: int) -> int:
        lower = 0
        upper = len(self._text) - 1

        while lower <= upper:
            index = lower + (upper - lower) // 2
            start = self.lines[index].start

            if position == start:
                return index

            if start > position:
                upper = index - 1
            else:
                lower = index + 1

        return lower + 1

    def __str__(self):
        return self._text

    def to_str(self, span: TextSpan) -> str:
        return self._text[span.start : span.end]

    def __len__(self):
        return len(self._text)

    def __getitem__(self, item):
        return self._text[item]

File: cpp/minsk/source/analysis/text/source.cpp
#include "minsk/analysis/text/source.hpp"
#include "minsk/analysis/text/span.hpp"

minsk::analysis::text::source_text::source_text(std::string &&text)
    : m_text(std::move(text)), m_lines(parse_lines(this, m_text)) {}

std::vector<minsk::analysis::text::text_line>
minsk::analysis::text::source_text::parse_lines(source_text *st,
                                                std::string_view text) {
  auto result = std::vector<text_line>{};
  int line_start = 0;
  int position = 0;
  while (position < text.size()) {
    auto line_break_width = get_line_break_width(text, position);
    if (line_break_width > 0) {
      add_line(&result, line_start, position, line_break_width);
      position += line_break_width;
      line_start = position;
    } else {
      position += 1;
    }
  }
  add_line(&result, line_start, position, 0);
  return result;
}

int minsk::analysis::text::source_text::get_line_break_width(
    std::string_view text, int i) {
  char c = text[i];
  char look = i + 1 == text.size() ? '\0' : text[i + 1];
  if (c == '\r' && look == '\n') {
    return 2;
  }
  if (c == '\r' || c == '\n') {
    return 1;
  }
  return 0;
}

void minsk::analysis::text::source_text::add_line(
    std::vector<text_line> *result, int line_start, int position,
    int line_break_width) {
  int length = position - line_start;
  int length_including_line_break = length + line_break_width;
  result->emplace_back(line_start, length, length_including_line_break);
}

minsk::analysis::text::source_text
minsk::analysis::text::source_text::from(std::string &&text) {
  return source_text{std::move(text)};
}

int minsk::analysis::text::source_text::get_line_index(int position) {
  int lower = 0;
  int upper = static_cast<int>(m_lines.size()) - 1;

  while (lower <= upper) {
    int index = lower + (upper - lower) / 2;
    int start = m_lines[index].start();

    if (position == start) {
      return index;
    }

    if (start > position) {
      upper = index - 1;
    } else {
      lower = index + 1;
    }
  }

  return lower - 1;
}

std::string_view minsk::analysis::text::source_text::text() const {
  return m_text;
}

const std::vector<minsk::analysis::text::text_line> &
minsk::analysis::text::source_text::lines() const {
  return m_lines;
}

std::string minsk::analysis::text::source_text::to_string(int start,
                                                          int length) const {
  return m_text.substr(start, length);
}

std::string
minsk::analysis::text::source_text::to_string(text_span span) const {
  return to_string(span.start(), span.length());
}

char minsk::analysis::text::source_text::operator[](int index) const {
  return m_text[index];
}

int minsk::analysis::text::source_text::length() const {
  return static_cast<int>(m_text.length());
}

std::ostream &operator<<(std::ostream &os,
                         const minsk::analysis::text::source_text &st) {
  return os << st.text();
}
